src/modules/commander/Commander.cpp

```diff
@@ -1041,15 +1028,196 @@ Commander::handle_command(vehicle_status_s *status_local, const vehicle_command_
+
+				} else if ((int)(cmd.param5) == 1) {
+					/* accelerometer calibration */
+					answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+					status_flags.condition_calibration_enabled = true;
+					_worker_thread.startTask(WorkerThread::Request::AccelCalibration);
+
+				} else if ((int)(cmd.param5) == 2) {
+					// board offset calibration
+					answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+					status_flags.condition_calibration_enabled = true;
+					_worker_thread.startTask(WorkerThread::Request::LevelCalibration);
+
+				} else if ((int)(cmd.param5) == 4) {
+					// accelerometer quick calibration
+					answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+					status_flags.condition_calibration_enabled = true;
+					_worker_thread.startTask(WorkerThread::Request::AccelCalibrationQuick);
+
+				} else if ((int)(cmd.param6) == 1 || (int)(cmd.param6) == 2) {
+					// TODO: param6 == 1 is deprecated, but we still accept it for a while (feb 2017)
+					/* airspeed calibration */
+					answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+					status_flags.condition_calibration_enabled = true;
+					_worker_thread.startTask(WorkerThread::Request::AirspeedCalibration);
+
+				} else if ((int)(cmd.param7) == 1) {
+					/* do esc calibration */
+					if (check_battery_disconnected(&mavlink_log_pub)) {
+						answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+						status_flags.condition_calibration_enabled = true;
+						armed.in_esc_calibration_mode = true;
+						_worker_thread.startTask(WorkerThread::Request::ESCCalibration);
+
+					} else {
+						answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_DENIED, command_ack_pub);
+					}
+
+				} else if ((int)(cmd.param4) == 0) {
+					/* RC calibration ended - have we been in one worth confirming? */
+					if (status_flags.rc_input_blocked) {
+						/* enable RC control input */
+						status_flags.rc_input_blocked = false;
+						mavlink_log_info(&mavlink_log_pub, "Calibration: Restoring RC input");
+					}
+
+					answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+
+				} else {
+					answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_UNSUPPORTED, command_ack_pub);
+				}
+			}
+
+			break;
+		}
+
+	case vehicle_command_s::VEHICLE_CMD_FIXED_MAG_CAL_YAW: {
+			// Magnetometer quick calibration using world magnetic model and known heading
+			if ((status.arming_state == vehicle_status_s::ARMING_STATE_ARMED)
+			    || (status.arming_state == vehicle_status_s::ARMING_STATE_SHUTDOWN)
+			    || _worker_thread.isBusy()) {
+
+				// reject if armed or shutting down
+				answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_TEMPORARILY_REJECTED, command_ack_pub);
+
+			} else {
+				answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
```

Comment:
Shouldn't the answer at least factor in the success or failure of starting the worker thread task?

Reply:
We could but I didn't see it worth doing, as it should be rather unlikely to fail (out of memory maybe).
I was also thinking about answering after the worker finished for at least some of those.

Reply:
Yes I'm thinking of out of memory on older boards (eg F4 + UAVCAN or too many mavlink instances). We currently handle that failure quite poorly, so it seems like an opportunity to at least inform the user what's happening rather than failing silently and probably hanging the QGC calibration screen.

We should either come back to it later or add something at another level.